File: app/src/main/java/mariuszbaleczny/compass/CompassActivity.java
package mariuszbaleczny.compass;

import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

public class CompassActivity extends AppCompatActivity {

    private boolean pressBackAgainToExit;
    private Toast compassToast;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_compass);
        showFragmentIfNotExist(CompassFragment.newInstance(), CompassFragment.FRAGMENT_TAG, false);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_calibrate) {
            showToastIfInvisible(getString(R.string.calibration_message));
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        if (getSupportFragmentManager().getBackStackEntryCount() != 0 || pressBackAgainToExit) {
            super.onBackPressed();
        }
        pressBackAgainToExit = true;
        showToastIfInvisible(getString(R.string.press_again_to_exit_toast));

        new Handler().postDelayed(() -> pressBackAgainToExit = false, Constants.ON_BACK_PRESS_DELAY_TIME);
    }

    private void showFragmentIfNotExist(final Fragment fragment, final String tag, final boolean addToBackStack) {
        final Fragment existingFragment = getSupportFragmentManager().findFragmentById(R.id.container);

        if (isNullOrEqualsTo(fragment, existingFragment)) {
            final FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
            if (addToBackStack) {
                ft.addToBackStack(null);
            }
            ft.replace(R.id.container, fragment, tag);
            ft.commit();
        }
    }

    private boolean isNullOrEqualsTo(Fragment fragment, Fragment existingFragment) {
        return (existingFragment == null || !fragment.getClass().equals(existingFragment.getClass()));
    }

    private void showToastIfInvisible(CharSequence text) {
        if (isToastNullOrInvisible()) {
            compassToast = Toast.makeText(getApplicationContext(), text, Toast.LENGTH_SHORT);
            compassToast.show();
        }
    }

    private boolean isToastNullOrInvisible() {
        return compassToast == null || compassToast.getView().getWindowVisibility() != View.VISIBLE;
    }
}

File: app/src/main/kotlin/mariuszbaleczny/compass/ui/activity/CompassActivity.kt
package mariuszbaleczny.compass.ui.activity

import android.os.Bundle
import android.os.Handler
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import mariuszbaleczny.compass.Constants
import mariuszbaleczny.compass.R
import mariuszbaleczny.compass.R.id
import mariuszbaleczny.compass.R.layout
import mariuszbaleczny.compass.R.string
import mariuszbaleczny.compass.dagger.DaggerCompassComponent
import mariuszbaleczny.compass.mvp.CompassMvp
import mariuszbaleczny.compass.ui.fragment.CompassFragment
import javax.inject.Inject

/**
 * Created by mariusz on 31.10.16.
 */
class CompassActivity : AppCompatActivity() {

    @Inject
    lateinit var compassView: CompassMvp.View

    private var toast: Toast? = null
    private var pressBackAgainToExit: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout.activity_compass)
        DaggerCompassComponent.builder().build().inject(this)
        showFragment(compassView.asFragment(), CompassFragment.TAG, false)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item?.itemId == id.action_calibrate) {
            showToastIfInvisible(string.calibration_message)
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        if (supportFragmentManager.backStackEntryCount != 0 || pressBackAgainToExit) {
            return super.onBackPressed()
        }
        pressBackAgainToExit = true
        showToastIfInvisible(string.press_again_to_exit_toast)
        Handler().postDelayed({ pressBackAgainToExit = false }, Constants.ON_BACK_PRESS_DELAY_TIME.toLong())
    }

    private fun showFragment(fragment: Fragment, tag: String, addToBackStack: Boolean) {
        val existingFragment = supportFragmentManager.findFragmentById(id.container)

        if (existingFragment == null || fragment.javaClass != existingFragment.javaClass) {
            val transaction = supportFragmentManager.beginTransaction()
            if (addToBackStack) {
                transaction.addToBackStack(null)
            }
            transaction.replace(id.container, fragment, tag)
            transaction.commit()
        }
    }

    private fun showToastIfInvisible(resId: Int) {
        if (!isToastVisible()) {
            toast = Toast.makeText(this, resId, Toast.LENGTH_SHORT)
            toast?.show()
        }
    }

    private fun isToastVisible(): Boolean = (toast?.view?.windowVisibility == View.VISIBLE)

}